docs(integration): update spec for encoded payload definition and update blob decoding (#1891)
Layr-Labs/eigenda: docs/spec/src/integration/spec/6-secure-integration.md

  - Implementation varies by requirement (e.g., key-value mapping for optimistic fault proofs)
- **Host:** Entity that provides preimage oracle responses

> > A encoded payload is an intermdeidate transformation form between the rollup payload and the EigenDA blob. Compared to payload, the encoded payload contains 

Reviewer: ```suggestion
> An encoded payload is an intermediate form in-between the transformation of rollup payload to EigenDA blob. Compared to rollup payload, the encoded payload contains 
```

### 1. RBN Recency Validation

This check enforces timing guarantees: once a cert lands in the batcher inbox, optimistic and zk rollup validators must have enough time to download the EigenDA blob.

We use fault proofs to motivate the need for a recency check.

![](../../assets/integration/recency-window-timeline.png)

> From the timeline above, EigenDA’s availability window must overlap the ~7-day challenge period so any honest party can detect faults and fetch the required data. Rollup derivation pipelines should reject certificates whose DA window began too far in the past. While a DA cert doesn’t record its signing or availability time, it does include cert.RBN, which is the L1 Reference Block Number chosen by the disperser to anchor the operator set and stakes. Because RBN is fixed before validators sign, it provides a proxy to bound how old the cert can be at inclusion, enabling a simple recency check.

Reviewer: technically challenge + resolution right?

Author: we only worry about the challenge period, I think it is generally acceptable to argument once you start the resolution process, you already have downloaded the blobs

Reviewer: that assumption makes sense

```
certL1InclusionBlock - cert.RBN <= RecencyWindowSize
```

> If the inequality fails, discard the cert. This also hardens security by preventing a disperser from choosing a very old RBN with materially different stakes (e.g., after withdrawals).

Reviewer 2: is it is it?? Does @ethenotethan speak like an LLM then? :D

> If the inequality fails, discard the cert. This also hardens security by preventing a disperser from choosing a very old RBN with materially different stakes (e.g., after withdrawals).

Author: But I though those are good words. materially good

> To give a concrete example with a rollup stack, optimism has a [sequencerWindow](https://docs.optimism.io/stack/rollup/derivation-pipeline#sequencer-window) which forces batches to land onchain in a timely fashion (12h). This filtering however, happens in the [BatchQueue](https://specs.optimism.io/protocol/derivation.html#batch-queue) stage of the derivation pipeline (DP). But because EigenDA blob derivation needs to take place right after [L1Retrieval](https://specs.optimism.io/protocol/derivation.html#l1-retrieval) and before [BatchQueue], we cannot use the OP's existing mechanism in [BatchQueue] with [sequencerWindow] to discard old DA certificate. To prevent this, we need the recencyWindow filtering to happen during the L1Retrieval stage of the DP.
>
> Despite its semantics being slightly different, sequencerWindow and recencyWindow are related concepts, and in order to not force another config change on op altda forks, we suggest using the same value as the `SequencerWindowSize` for the `RecencyWindowSize`, namely 12h.

> To ensure the ~7-day challenge window overlaps EigenDA availability, we assume the L2 state proposer posts to L1 frequently so the gap between (a) the L1 inclusion of the cert in the batcher inbox and (b) the L1 inclusion of the corresponding L2 state update stays small. Call this gap L2StatePostingPeriod. To leave validators enough time to detect and download the data, require: L2StatePostingPeriod + RecencyWindowSize < ~7 days.

Author: Don't think it is me. 

I don't intend to make it a check, because we currently don't have a programmatic way to enforce it. It is more of a recommendation or reasoning. But if there is someone who stores all blobs, such assumption is not needed. 

This is already the case for L2, but there is assumption of archival node that makes people not worrying about it.

I remember you and someone from L2beat argued about it before. I think he is convinced that the assumption of archival node can be removed. 

To clarify further,  L2StatePostingPeriod requires just one party to publish the L2 claim on L1 within the time window.

![](../../assets/integration/cert-rbn-recency-window.png)

In the diagram, the top row shows L1 blocks every 12 s; the smaller squares are L2 blocks every 2 s. Yellow labels mark key artifacts across the batching pipeline: batches → channel → EigenDA blob. Dispersal completes between t=12 s and t=24 s. The resulting certificate has RBN equal to the L1 block at t=0 (two L1 blocks earlier). The cert is then submitted to L1 at t=24 s. Green annotations show the generalized L2→L1 submission, with batches posted to the adjacent L1 block.

### 2. Cert Validation

> Cert validation is done inside the EigenDACertVerifier contract, which EigenDA deploys as-is, but is also available for rollups to modify and deploy on their own. Specifically, [checkDACert](https://github.com/Layr-Labs/eigenda/blob/2414ed6f11bd28bc631eab4da3d6b576645801b0/contracts/src/periphery/cert/EigenDACertVerifier.sol#L46-L56) is the entry point for validation. This could either be called during a normal eth transaction (either for pessimistic “bridging” like EigenDA V1 used to do, or when uploading a Blob Field Element to a one-step-proof’s [preimage contract](https://specs.optimism.io/fault-proof/index.html#pre-image-oracle)), or be zk proven using a library like [Steel](https://docs.beboundless.xyz/developers/steel/what-is-steel) and [Sp1CC](https://succinctlabs.github.io/sp1-contract-call/).

Reviewer: can we link to the cert verifier router overview from here?

Reviewer: Also should we define Blob Field Element in our glossary? https://layr-labs.github.io/eigenda/integration/spec/3-data-structs.html?highlight=blob%20field%20element#blob

3. [verify](https://github.com/Layr-Labs/eigenda/blob/3e670ff3dbd3a0a3f63b51e40544f528ac923b78/contracts/src/periphery/cert/legacy/v2/EigenDACertVerificationV2Lib.sol#L198-L218) blob security params (blob_params + security thresholds)
4. [verify](https://github.com/Layr-Labs/eigenda/blob/3e670ff3dbd3a0a3f63b51e40544f528ac923b78/contracts/src/periphery/cert/legacy/v2/EigenDACertVerificationV2Lib.sol#L259-L279) each quorum part of the blob_header has met its threshold

### 3. Downloading and Decoding an Encoded Payload

#### Downloading an Encoded Payload

> The preimage oracle served encoded payload. When the EigenDA blob derivation queries the preimage oracle for the encoded payload corresponding to a DA cert, the preimage oracle (i.e. the preimage request module of the EigenDA proxy) downloads the EigenDA blob from relay or directly from EigenDA operators.

Reviewer: let's embed a link to the encoded payload definition

> The preimage oracle served encoded payload. When the EigenDA blob derivation queries the preimage oracle for the encoded payload corresponding to a DA cert, the preimage oracle (i.e. the preimage request module of the EigenDA proxy) downloads the EigenDA blob from relay or directly from EigenDA operators.

Reviewer: ```suggestion
The preimage oracle serves the EigenDA blob in the [encoded payload form](link). When the EigenDA blob derivation queries the preimage oracle for the encoded payload corresponding to a DA cert, the preimage oracle (i.e. the preimage request module of the EigenDA proxy) downloads the EigenDA blob from relay or directly from EigenDA operators.
```

Reviewer: also not sure about this system flow assumption:
```
 When the EigenDA blob derivation queries the preimage oracle for the encoded payload corresponding to a DA cert, the preimage oracle (i.e. the preimage request module of the EigenDA proxy) downloads the EigenDA blob from relay or directly from EigenDA operators.
 ```
 
 assuming this spec is generalized for optimistic rollups, we may wanna change the description to reflect that preimage fetching can vary (e.g, AOT like in Arbitrum)

Author: @ethenotethan how about this, https://github.com/Layr-Labs/eigenda/pull/1891/commits/d8cae9732f7a8aa3e59c50bdfe926d1c4d0c8e67

> The preimage oracle served encoded payload. When the EigenDA blob derivation queries the preimage oracle for the encoded payload corresponding to a DA cert, the preimage oracle (i.e. the preimage request module of the EigenDA proxy) downloads the EigenDA blob from relay or directly from EigenDA operators.

Author: Actually from Ethen's comment, we should have the preimage oracle to encompass all data source, so payload retriever isn't right. 

https://github.com/Layr-Labs/eigenda/pull/1891#discussion_r2285086492

If verification fails, it discards the blob and retries with other sources until a valid one is found. Once verified, it returns the encoded payload to the derivation step.
The preimage oracle performs checks on the blob against the KZG commitment from the DA cert. 

> If verification fails, it discards the blob and retries with other sources until a valid one is found. Once verified, it returns the encoded payload to the derivation step.

Reviewer: looks like this is duplicated with line 166

> A rollup may apply an FFT on the blob to obtain its encoded payload, or use the blob directly as the encoded payload. 
> This depends on if there is an inverse FFT is taken on the encoded payload in the dispersal path. Taking IFFT on the dispersal path lets the rollup open points on bytes using parts of the payload. Both Arbitrum Nitro and OP (optimistic or ZK) apply IFFT. The encoded payload always live in the same domain (i.e. without any data transformation) as the payload. It is formed by adding the encoded payload header and interleaving 0s to make every 32bytes a valid field element, the padding 0s at the end to a power of two number of field elements (each 32 bytes).
> The simplest kzg check recomputes the commitment from the SRS points and compare it. An alternative is verifying a point opening via Fiat–Shamir (see this issue).

#### Decoding an Encoded Payload

After verification, EigenDA blob derivation decodes the encoded payload to the original rollup payload. If any check fails, discard the blob returned from the preimage oracle. The procedure:

- checkLenInvariant

>   - Encoded payload size ≥ size of encoded payload header.

Reviewer: think we should give a formal definition for `encoded payload header` - brought this up in the glossary section earlier in review

Author: https://github.com/Layr-Labs/eigenda/pull/1891/commits/8f7252fc4757aa22ad36f104d303969a6138d28f#diff-60c7fb1452706db7b55c417746564caa44b3180eb2e8ac4e50c0ebfecfa23bd7R176

>   - Total length of encoded payload perfectly is a multiple of 32 bytes.

Reviewer: ```suggestion
  - Length of encoded payload is a multiple of 32 bytes (l.e, `len(encoded_payload)%32==0`)
```

  - Encoded payload contains a power-of-two number of 32-byte field elements (valid sizes: 32, 64, 128, 256, …). See client [implementation](https://github.com/Layr-Labs/eigenda/blob/57ed95ce77a57c53341cad10233ca2f29b29c0f5/api/clients/v2/coretypes/encoded_payload.go#L152).
- decodeHeader: (first 32-byte field element)
  - Encoded payload size ≥ size of encoded payload header.
  - First byte is 0x00 so the first 32 bytes form a valid field element.
  - Encoding version is known (currently 0x00).
  - Returns the claimed original rollup payload size.
- decodePayload
  - Remove internal padding (drop the first byte of each 32-byte word).
  - Decoded size must be ≥ the claimed length.

> The EigenDA protocol enforces blob length > 0 (see [implementation](https://github.com/Layr-Labs/eigenda/blob/57ed95ce77a57c53341cad10233ca2f29b29c0f5/node/grpc/server_v2.go#L127)).

Proxy behavior. The EigenDA proxy can return either the encoded payload or the decoded rollup payload based on GET parameters:

>   - With `?return_encoded_payload=true` or `?return_encoded_payload=1`, it only performs the preimage (KZG) check and returns the encoded payload useful when integrating with proof systems to control the data transformation.

Reviewer: what is the preimage kzg check?

Author: https://github.com/Layr-Labs/eigenda/pull/1891/commits/8f7252fc4757aa22ad36f104d303969a6138d28f#diff-60c7fb1452706db7b55c417746564caa44b3180eb2e8ac4e50c0ebfecfa23bd7R193
